google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/AbstractJdbcResultSet.java

```diff
@@ -0,0 +1,627 @@
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner.jdbc;
+
+import java.io.InputStream;
+import java.io.Reader;
+import java.math.BigDecimal;
+import java.sql.Array;
+import java.sql.Blob;
+import java.sql.Clob;
+import java.sql.Date;
+import java.sql.NClob;
+import java.sql.Ref;
+import java.sql.ResultSet;
+import java.sql.RowId;
+import java.sql.SQLException;
+import java.sql.SQLFeatureNotSupportedException;
+import java.sql.SQLWarning;
+import java.sql.SQLXML;
+import java.sql.Time;
+import java.sql.Timestamp;
+
+/** Base class for Cloud Spanner {@link ResultSet}s. */
+abstract class AbstractJdbcResultSet extends AbstractJdbcWrapper implements ResultSet {
+  /** The underlying Cloud Spanner {@link com.google.cloud.spanner.ResultSet}. */
+  final com.google.cloud.spanner.ResultSet spanner;
+
+  AbstractJdbcResultSet(com.google.cloud.spanner.ResultSet spanner) {
+    this.spanner = spanner;
+  }
+
+  @Override
+  public SQLWarning getWarnings() throws SQLException {
+    return null;
+  }
+
+  @Override
+  public void clearWarnings() throws SQLException {}
+
+  @Override
+  public String getCursorName() throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  @Override
+  public boolean isLast() throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  @Override
+  public void beforeFirst() throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  @Override
+  public void afterLast() throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  @Override
+  public boolean first() throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  @Override
+  public boolean last() throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  @Override
+  public boolean absolute(int row) throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  @Override
+  public boolean relative(int rows) throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  @Override
+  public boolean previous() throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  @Override
+  public void setFetchDirection(int direction) throws SQLException {
+    JdbcPreconditions.checkArgument(direction == FETCH_FORWARD, direction);
+  }
+
+  @Override
+  public int getFetchDirection() throws SQLException {
+    return FETCH_FORWARD;
+  }
+
+  @Override
+  public void setFetchSize(int rows) throws SQLException {
+    // no-op
+  }
+
+  @Override
+  public int getFetchSize() throws SQLException {
+    return 0;
+  }
+
+  @Override
+  public int getType() throws SQLException {
+    return TYPE_FORWARD_ONLY;
+  }
+
+  @Override
+  public int getConcurrency() throws SQLException {
+    return CONCUR_READ_ONLY;
+  }
+
+  @Override
+  public boolean rowUpdated() throws SQLException {
+    return false;
```

Comment:
do any of these have meaningful values? If no, should we just throw unsupported for all of them?

Reply:
There's something to say for both throwing `SQLFeatureNotSupportedException` and for returning `false` in these cases. The JDBC specification does not specify exactly what a driver should do with these methods when the `updateXYZ` are not supported, only the other way around: When `updateXYZ` is supported, then these methods must also be implemented. In general, I've chosen the following generic approach for all methods in the JDBC driver where the implementation is optional:
* Return a sensible (constant) value when the driver can determine one, even when the other methods that would normally be used to change the state that is returned by the method are not supported.
* Throw a `SQLFeatureNotSupportedException` when no reasonable (constant) value can be returned, or when the method should alter the state of the object and it is not supported by the driver or Cloud Spanner.

In this specific case we know that the row has not been updated, because we do not support the update methods. This is also reflected in the return value of the `DatabaseMetaData#updatesAreDetected(int)` method, so a user should be able to know beforehand that calling these methods will not have any effect.

The reasoning behind the above generic approach is that the driver can also be used with generic third-party tools like for example DBeaver and with all sorts of generic Java frameworks. As both we and the end user have no influence on how these tools use the driver, it will give a better user experience when we are little bit lenient when possible, rather than throwing exceptions for everything that is not supported, even when the tool in question could have determined that calling the methods is not supported.